File: php/Environment.php
<?php
// Environment.php - 运行时环境

class Environment {
    private $context;          // 全局上下文（如 payload）
    private $variables = [];   // 变量栈（作用域）
    private $functions = [];   // 函数表: [name][argCount] = FunctionDefNode|callable
    private $controls = [];    // 控件表: [name] => callable
    private $scopeStack = [];
    private $lastRunFuncResult = null; // 最近一次 @RunFunc 的结果

    public function __construct($context = []) {
        $this->context = $context;
        $this->pushScope();
        foreach ($context as $key => $value) {
            $this->setVariable($key, $value);
        }
    }

    // ---------- 变量 ----------
    public function pushScope() {
        $this->scopeStack[] = [];
    }
    public function popScope() {
        array_pop($this->scopeStack);
    }
    public function setVariable($name, $value) {
        // 如果变量在父作用域已存在，修改父作用域中的值
        for ($i = count($this->scopeStack) - 2; $i >= 0; $i--) {
            if (isset($this->scopeStack[$i][$name])) {
                $this->scopeStack[$i][$name] = $value;
                return;
            }
        }
        // 否则在当前作用域创建
        $this->scopeStack[count($this->scopeStack)-1][$name] = $value;
    }
    public function getVariable($name) {
        for ($i = count($this->scopeStack) - 1; $i >= 0; $i--) {
            if (isset($this->scopeStack[$i][$name])) {
                return $this->scopeStack[$i][$name];
            }
        }
        // 回退到上下文
        if (isset($this->context[$name])) {
            return $this->context[$name];
        }
        throw new Exception("Undefined variable: $name");
    }

    // ---------- 函数 ----------
    public function registerFunction($name, $def, $argCount) {
        if (!isset($this->functions[$name])) {
            $this->functions[$name] = [];
        }
        $this->functions[$name][$argCount] = $def;
    }
    public function hasFunction($name, $argCount) {
        if (!is_string($name)) {
            throw new Exception("Function name must be a string, got " . gettype($name));
        }
        return isset($this->functions[$name]) && isset($this->functions[$name][$argCount]);
    }
    public function callFunction($name, $args) {
        if (!is_string($name)) {
            throw new Exception("Function name must be a string, got " . gettype($name));
        }
        $argCount = count($args);
        if (!$this->hasFunction($name, $argCount)) {
            throw new Exception("Function $name with $argCount arguments not found");
        }
        $def = $this->functions[$name][$argCount];
        if ($def instanceof FunctionDefNode) {
            // 创建新的作用域，绑定参数
            $this->pushScope();
            // 设置 Param 变量，包含参数数量和参数列表
            // 通过 Param:type 格式时，参数名是类型字符串（如 'any'）
            $params = $def->params;
            if (count($params) === 1 && !$this->isVariableParam($params[0])) {
                // Param:any 格式 — 将整个参数值作为 Param
                $this->setVariable('Param', $args[0] ?? null);
            } else {
                // Param:$a,$b 格式 — Param 为参数信息对象
                $this->setVariable('Param', [
                    'quantity' => $argCount,
                    'args' => $args
                ]);
            }
            foreach ($def->params as $idx => $paramName) {
                $this->setVariable($paramName, $args[$idx] ?? null);
            }
            $result = $def->body->execute($this);
            $this->popScope();
            return $result;
        } elseif (is_callable($def)) {
            return call_user_func_array($def, $args);
        }
        throw new Exception("Invalid function definition");
    }

    // ---------- 控件 ----------
    public function registerControl($name, callable $callable) {
        $this->controls[$name] = $callable;
    }
    public function hasControl($name) {
        return isset($this->controls[$name]);
    }
    public function callControl($name, $args) {
        if (!$this->hasControl($name)) {
            throw new Exception("Control $name not registered");
        }
        return call_user_func_array($this->controls[$name], $args);
    }

    // 获取上下文
    public function getContext() {
        return $this->context;
    }
    public function setContext($newContext) {
        $this->context = $newContext;
    }

    // 重置作用域栈（用于 @EventRestart 后重新开始执行）
    public function resetScopes() {
        $this->scopeStack = [];
        $this->pushScope();
        if (is_array($this->context)) {
            foreach ($this->context as $key => $value) {
                $this->setVariable($key, $value);
            }
        } else {
            // 非数组 context（如单个值），设置为 payload 变量
            $this->setVariable('payload', $this->context);
        }
    }

    // RunFunc 结果存取
    public function setRunFuncResult($result) {
        $this->lastRunFuncResult = $result;
    }
    public function getRunFuncResult() {
        return $this->lastRunFuncResult;
    }

    // 判断参数名是否为变量名（非类型关键字）
    private function isVariableParam($name) {
        $typeKeywords = ['any', 'bool', 'string', 'int', 'number', 'float', 'double', 'array', 'object', 'callable', 'void', 'null', 'mixed'];
        return !in_array($name, $typeKeywords);
    }
}

File: python/environment.py
"""CCHQCode Runtime - 运行时环境"""

from typing import Any, List, Optional

from .types import ControlSignal, EventRestartSignal

TYPE_KEYWORDS = {
    'any', 'bool', 'string', 'int', 'number',
    'float', 'double', 'array', 'object',
    'callable', 'void', 'null', 'mixed',
}


class Environment:
    def __init__(self, context: Any = None):
        self.context = context if context is not None else {}
        self.scope_stack: List[dict] = []
        self.functions: dict = {}
        self.controls: dict = {}
        self._last_run_func_result = None

        self.push_scope()
        if isinstance(self.context, dict):
            for k, v in self.context.items():
                self.set_variable(k, v)

    # ---------- 作用域 ----------
    def push_scope(self):
        self.scope_stack.append({})

    def pop_scope(self):
        self.scope_stack.pop()

    def set_variable(self, name: str, value: Any):
        # 如果变量在父作用域已存在，修改父作用域中的值
        for i in range(len(self.scope_stack) - 2, -1, -1):
            if name in self.scope_stack[i]:
                self.scope_stack[i][name] = value
                return
        # 否则在当前作用域创建
        self.scope_stack[-1][name] = value

    def get_variable(self, name: str) -> Any:
        for scope in reversed(self.scope_stack):
            if name in scope:
                return scope[name]
        # 回退到上下文
        if isinstance(self.context, dict) and name in self.context:
            return self.context[name]
        raise ValueError(f"Undefined variable: {name}")

    # ---------- 函数 ----------
    def register_function(self, name: str, def_, arg_count: int):
        if name not in self.functions:
            self.functions[name] = {}
        self.functions[name][arg_count] = def_

    def has_function(self, name: str, arg_count: int) -> bool:
        return name in self.functions and arg_count in self.functions[name]

    def call_function(self, name: str, args: List[Any]) -> Any:
        if not isinstance(name, str):
            raise TypeError(f"Function name must be a string, got {type(name)}")
        arg_count = len(args)
        if not self.has_function(name, arg_count):
            raise ValueError(f"Function {name} with {arg_count} arguments not found")

        def_ = self.functions[name][arg_count]
        # 鸭子类型检查: 有 body 和 params 就是函数定义节点
        if hasattr(def_, 'body') and hasattr(def_, 'params'):
            self.push_scope()
            params = def_.params
            is_type_param = len(params) == 1 and params[0] in TYPE_KEYWORDS

            if is_type_param:
                self.set_variable('Param', args[0] if args else None)
            else:
                self.set_variable('Param', {
                    'quantity': arg_count,
                    'args': list(args),
                })
            for idx, pname in enumerate(params):
                self.set_variable(pname, args[idx] if idx < len(args) else None)

            try:
                result = def_.body.execute(self)
                self.pop_scope()
                return result
            except EventRestartSignal as e:
                self.pop_scope()
                return self.call_function(name, [e.new_payload])
        elif callable(def_):
            return def_(*args)

        raise ValueError("Invalid function definition")

    # ---------- 控件 ----------
    def register_control(self, name: str, callable_):
        self.controls[name] = callable_

    def has_control(self, name: str) -> bool:
        return name in self.controls

    def call_control(self, name: str, args: List[Any]) -> Any:
        if not self.has_control(name):
            raise ValueError(f"Control {name} not registered")
        return self.controls[name](*args)

    # ---------- 上下文 ----------
    def get_context(self) -> Any:
        return self.context

    def set_context(self, ctx: Any):
        self.context = ctx

    def reset_scopes(self):
        self.scope_stack = []
        self.push_scope()
        if isinstance(self.context, dict):
            for k, v in self.context.items():
                self.set_variable(k, v)
        else:
            self.set_variable('payload', self.context)

    # ---------- RunFunc 结果 ----------
    def set_run_func_result(self, result: Any):
        self._last_run_func_result = result

    def get_run_func_result(self) -> Any:
        return self._last_run_func_result
